### PR TITLE
NEW profiler decorator

### PR DESCRIPTION
**Pull request by viniciusdc**
This should enable us to inspect more functions with more freedom.

### CODE REVIEW

**Comment by CJ-Wright** [comment]
Can you implement this on the current jobs?

**Comment by viniciusdc** [comment]
> Can you implement this on the current jobs?

Yes, it is possible.

**Comment by viniciusdc** [comment]
I will implement this on the `update_prs` profiler stuff, this way I can test it in a better way

**Comment by viniciusdc** [comment]
There is a problem with using `function_name = f.__name__` to get the process name (for the sub-folders), cause we are actuly using it on the `main` fucntions so... it is gonna be a mess. I will replace it with:
```python
# process name -- aka profiler sub-folder
process_name = os.path.basename(__file__)
# check dir
os.makedirs(f"profiler/{process_name}", exist_ok=True)
```

**Comment by viniciusdc** [comment]
Local test passed

**Comment by viniciusdc** [comment]
Also, after #1111 works, we could rethink this #1085, right ?

**Comment by CJ-Wright** [comment]
yes

File: conda_forge_tick/profiler.py [hunk 1]
```diff
@@ -9,25 +9,26 @@ def profiling():
...
-            # function name -- aka profiler sub-folder
-            function_name = f.__name__
-            # check dir
-            os.makedirs(f"profiler/{function_name}", exist_ok=True)
-            # save stats into file
-            prof.dump_stats(f"profiler/{function_name}/{current_time}")
-
-            return  # out_result
+            with cProfile.Profile() as prof:
+                prof = cProfile.Profile()
```
Comment (CJ-Wright) [inline conda_forge_tick/profiler.py:13]: why is this line here?

Comment (viniciusdc) [reply conda_forge_tick/profiler.py:13]: ops

Comment (viniciusdc) [reply conda_forge_tick/profiler.py]: done with the last commit

File: conda_forge_tick/profiler.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+from datetime import datetime
+
+
+# profiling decorator
+def profiling():
+    def _profiling(f):
+        @wraps(f)
+        def __profiling(*rgs, **kwargs):
+            with cProfile.Profile() as prof:
+                prof.enable()
```
Comment (CJ-Wright) [inline conda_forge_tick/profiler.py:13]: Do you still need the enable and disable?

Comment (viniciusdc) [reply conda_forge_tick/profiler.py:13]: not anymore, but I will read again the doc about `cProfile.Profile()` to check if there is any problem reggarding that (like the need of `prof.disable`).

Comment (viniciusdc) [reply conda_forge_tick/profiler.py:13]: Yup, I've checked [here](https://bugs.python.org/issue29235) an it seems to be unnecessary (redundant and wrong) the usage of the enable and disable in this case. Thanks for the remind CJ.

Comment (viniciusdc) [reply conda_forge_tick/profiler.py:13]: tested it locally, seams to be working fine. I will update the PR

**Comment by viniciusdc** [comment]
As [commented](https://github.com/regro/cf-scripts/pull/1111#issuecomment-674887491) by @CJ-Wright there is no need for two PR's with the same goal, then I will be closing #1111 as this PR now comprehends those changes too (in this case as a test for the new `@profiling()`)

File: conda_forge_tick/profiler.py (hunk 2 above)
Comment (viniciusdc) [reply conda_forge_tick/profiler.py]: Done

**Comment by viniciusdc** [comment]
> Can you implement this on the current jobs?

Once we test it with the `update_prs` I can insert the on the other process

**Comment by viniciusdc** [comment]
Once tested and working well, we should go to #1085

**Comment by CJ-Wright** [comment]
Is this decorator being applied to the auto tick runs as well?

**Comment by viniciusdc** [comment]
> Is this decorator being applied to the auto tick runs as well?

not yet

**Comment by viniciusdc** [comment]
I want to confirm that it will work as expected, here it goes well but there is always an unexpected thing that could happen

**Comment by viniciusdc** [comment]
If it works, I will update the autotick with `@profiling()` and work on the limit stuff (there is so much information there, only consuming space)

**Comment by CJ-Wright** [comment]
Do you need this PR merged first?

**Comment by viniciusdc** [comment]
> Do you need this PR merged first?

Yes, if possible. The order should be something like:
1. This PR with the `profile.py` and the `update_prs` (as a test);
2. Update `auto_tick` profiler using the `@profiling()`;
3. Remove old instances of profiled files

**Comment by CJ-Wright** [comment]
It would be good if all the changes related to the profiler decorator could be done in one PR.

**Comment by viniciusdc** [comment]
> It would be good if all the changes related to the profiler decorator could be done in one PR.

For that, I must figure out a way to remove the oldest files from `profiler/*`, do you have any idea ?

**Comment by CJ-Wright** [comment]
I would leave that portion out for the moment, since formally it isn't about the decorator, also that requires manual changes to the database.

**Comment by viniciusdc** [comment]
> I would leave that portion out for the moment, since formally it isn't about the decorator, also that requires manual changes to the database.

yup, Y 'r right. So for the profiling that's all here. Unless there is other part of the bot I should add the decorator.

**Comment by CJ-Wright** [comment]
You could put it on make graph, that might be interesting.
